modules/currency-exchange/src/lib.rs

```diff
@@ -0,0 +1,403 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::Encode;
+use frame_support::{decl_error, decl_module, decl_storage, ensure, fail, Parameter};
+use primitives::exchange::{
+	Airdrop, CurrencyConverter, Error as ExchangeError, MaybeLockFundsTransaction, RecipientsMap,
+};
+use sp_runtime::DispatchResult;
+use sp_std::vec::Vec;
+
+/// Peer blockhain interface.
+pub trait Blockchain {
+	/// Block hash type.
+	type BlockHash: Parameter;
+	/// Transaction type.
+	type Transaction: Parameter;
+	/// Transaction inclusion proof type.
+	type TransactionInclusionProof: Parameter;
```

Comment:
I'd convert the `Blockchain` trait into something more generic, like so:
```
struct RawTransaction(Vec<u8>);
trait TransactionInclusionProof {
  type BlockHash;
  type Header;

  fn transaction(&self) -> &RawTransaction;

  fn block_hash(&self) -> &Self::BlockHash;

  fn verify(&self, header: Header) -> bool;
}
```
and then implement it for a tuple (or a struct like this):
```
(H256 /* hash */, usize /* index */, Vec<RawTransaction> /* list of transactions */)
```

This does not require sending the transaction body twice in the payload and in the future will allow us to use different (more minimal) proof formats without changing much of the code (we can even support both formats if needed).
The verification logic stays close to the proof data as well, which is something that I like too. We can keep the transaction parsing separate though. 

Not super strong on this though, since I feel that it might not be generic enough and might actually complicate some other stuff (namely UX) - so open to discussion.